File: CPP/BOJ-S1-1629-Multiplication.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
ll a, b, c;
ll go(ll a, ll b)
{
  if (b == 1)
    return a % c;

  ll ret = go(a, b / 2);
  ret = (ret * ret) % c;

  // b가 홀수일 경우
  if (b % 2)
    ret = (ret * a) % c;
  return ret;
}

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> a >> b >> c;

  cout << go(a, b) << '\n';

  return 0;
}

File: JS/BOJ-S1-1629-Multiplication.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split(' ')
  .map((v) => v.trimEnd());

const go = (a, b, c) => {
  if (b == 1) return a % c;
  let ret = go(a, b / BigInt(2), c);
  ret = (ret * ret) % c;
  if (b % BigInt(2)) return (ret * a) % c;
  return ret;
};

let output = [];
const solution = (input) => {
  const [a, b, c] = input.map(BigInt);
  output.push(go(a, b, c));
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );
